### PR TITLE
Dynamic verification of Channels

### LINKED ISSUE #88: Channels should verify their usage at runtime

### PR DESCRIPTION
**Pull request by irvall**
Essentially, the channel gets the recursive type given as a parameter and parse it to a statemachine it will then progress in as we call send/recv/choose/offer on it. Unfortunately, stuff hangs due to no dual part to i.e. send something upon receiving.

closes #88

### CODE REVIEW

File: src/channel.py [hunk 1]
```diff
@@ -1,19 +1,27 @@
...
+import statemachine
+from statemachine import Action
+from check import typecheck_file
+
 
-T = TypeVar('T')
 
+T = TypeVar('T')
 
 class Channel(Generic[T]):
```
Comment (fremartini) [inline src/channel.py:17]: remove the 'local mode' concept. We dont need threading

File: src/check.py [hunk 2]
```diff
@@ -11,20 +11,24 @@
...
+visited_files = {}
+
 class TypeChecker(NodeVisitor):
 
     def __init__(self, tree, inside_class=False) -> None:
         self.inside_class = inside_class
         self.function_queue = {}
         self.functions_that_alter_channels = {}
         self.subst_var = {}
+        self.channels = {}
```
Comment (fremartini) [inline src/check.py:23]: check usage

**Review by fremartini — CHANGES_REQUESTED** [review]

File: src/channel.py (hunk 1 above)
Comment (irvall) [reply src/channel.py:17]: That, you will do, young padawan

File: src/channel.py [hunk 3]
```diff
@@ -106,3 +147,6 @@ def _trace(ex: Exception) -> None:
 
 def _exit() -> None:
     sys.exit(0)
+
+def session_channel(s1: Session, s2: Session) -> Tuple[Channel, Channel]:
```
[the remark this answers is not in the document]
Comment (irvall) [reply src/channel.py]: Removed

File: src/session.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+from types import GenericAlias
+from unicodedata import name
+import sessiontype
+from sessiontype import Choose, Send, Recv, End, Offer, Choose, Label
+from typing import ForwardRef
+
+class Session():
```
[the remark this answers is not in the document]
Comment (irvall) [reply src/session.py:7]: Done

File: src/sessiontype.py [hunk 5]
```diff
@@ -37,6 +52,17 @@ class End(SessionType):
 class SessionException(TypeError):
     ...
 
+def has_dual(typ):
```
[the remark this answers is not in the document]
Comment (irvall) [reply src/sessiontype.py]: No, removed

**Review by fremartini — APPROVED** [review]

**Pull request merged**